src/MSBuildBinLogQuery/Token/WarningToken.cs

```diff
@@ -1,27 +1,24 @@
-using System;
-using System.Diagnostics.CodeAnalysis;
-
 namespace Microsoft.Build.Logging.Query.Token
 {
-    public sealed class WarningToken : Token, IEquatable<WarningToken>
+    public class WarningToken : Token
     {
-        public WarningToken() : base()
+        public static WarningToken Instance
         {
-        }
+            get
+            {
+                if (_instance == null)
+                {
+                    _instance = new WarningToken();
+                }
```

Comment:
I would suggest this:

``` C#
public static WarningToken Instance { get; } = new WarningToken();
```

The way you have it has a race condition if called on multiple threads.  This will be faster to access too with no branching. I don't think the laziness is necessary since the operation is so cheap. Above will allow the runtime to run the WarningToken when it sees fit as long as it is before the first use of static field. In practice, all of the constructors will run once during JIT compilation of the scanner method that uses Instance, and then the native code can just retrieve the static instance without a branch.